### PR TITLE
Reproject implementation

### LINKED ISSUE #429: Initial NDCube.resample Method
### Description

Create a method on `NDCube` to resample the data grid based on a new WCS.
* Should take a target APE-14-compliant WCS object from the user to which the data should be reprojected.  Initially this WCS should describe all axes of the cube.  Future versions should handle invariant axes, but not as part of this issue.
* Verify the target WCS is compatible with `NDCube`'s WCS.  Use the function outlined in #428
* Use `reproject.reproject_interp` to reproject `NDCube.data`.
* Return a new `NDCube` using the target WCS, the projected data array, the `NDCube.meta` and the `global_coords`.  Drop the `uncertainty`, `mask` and `extra_coords` information.  (Future versions can handle these cases.)

### PR DESCRIPTION
**Pull request by adwaitbhope**
### Description

Fixes #429.

For now, this is basically a wrapper around reproject, will add tests later.
Trying to figure out a way to compute `shape_out` from available data.

### CODE REVIEW

File: ndcube/ndcube.py
Comment (DanRyanIrish) [inline ndcube/ndcube.py:679]: Does `meta` and `global_coords` have to be deepcopies here?

File: ndcube/ndcube.py
Comment (DanRyanIrish) [inline ndcube/ndcube.py:681]: The check of `return_footprint` needs to be made here.  According to docstring, only when `return_footprint=True` will `footprint` be returned.

File: ndcube/ndcube.py
Comment (DanRyanIrish) [inline ndcube/ndcube.py:673]: Handling of `return_footprints` needs to be done here.

**Comment by DanRyanIrish** [comment]
I think this is a great start @adwaitbhope.  When you have a simple test confirming the ouput, we can talk about how next to proceed.

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:681]: Right, missed that. In that case, should we handle this similar to reproject, or should we return `resampled_cube, None` if `return_footprint=False`?

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:679]: Well, I thought it would be better to not have any references to the original cube's attributes.

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py]: I've added a commit that handles it like reproject for now.

File: ndcube/ndcube.py
Comment (Cadair) [reply ndcube/ndcube.py:685]: yeah probably wise.

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:671]: You should add a test which raises this exception.

**Review by Cadair — CHANGES_REQUESTED** [review]
Looking good @adwaitbhope just a few pretty minor comments :smile:

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py]: Oops, sorry. Copied parts of the docstring from elsewhere, fixed it now.

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:667]: Added

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:687]: Yeah, this does look much cleaner. I'll make this change now, perhaps we could make a call about the footprint later?
Also, if we're going to calculate the footprint anyway, would it be better to return it as well?

Comment (Cadair) [reply ndcube/ndcube.py:687]: yeah the more I think about it exposing the option like this is probably the best way.

Comment (adwaitbhope) [reply ndcube/ndcube.py:687]: Maybe we could set the default argument to `False`?

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py]: In case both are available, should we treat `shape_out` as an override or go with `.pixel_shape` anyway?

Comment (Cadair) [reply ndcube/ndcube.py]: I would take `shape_out` to be an override. Explicit things should generally always override implicit things.

File: ndcube/ndcube.py
Comment (Cadair) [reply ndcube/ndcube.py]: sounds good.

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:635]:
```suggestion
        Reprojects this NDCube to the coordinates described by another WCS object.
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:640]:
```suggestion
            The WCS object on which the ``NDCube`` is to be reprojected.
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:644]:
```suggestion
            The shape of the output data. If not specified, `~astropy.wcs.wcsapi.BaseLowLevelWCS.pixel_shape` attribute (if available)
            from the low level API of the ``target_wcs`` is used.
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:646]:
```suggestion
        order: `int` or `str`
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:650]:
```suggestion
        output_array: `numpy.ndarray`, optional
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:660]:
```suggestion
            A new resultant NDCube object, the supplied ``target_wcs`` will be the ``.wcs`` attribute of the output ``NDCube``.
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:669]:
```suggestion
        This method doesn't support handling of the ``mask``, ``extra_coords``, and ``uncertainty`` attributes yet.
        However, ``meta`` and ``global_coords`` are copied to the output ``NDCube``.
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:673]:
```suggestion
            raise Exception('Given target_wcs is not compatible with this NDCube, the physical types do not match.')
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:679]: You need to do the `is not None` any time you are comparing to a thing which could be an iterable.
```suggestion
            if hasattr(low_level_target_wcs, 'pixel_shape') and low_level_target_wcs.pixel_shape is not None:
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:683]:
```suggestion
                raise Exception("shape_out must be specified if target_wcs's low level API "
                                "does not have the pixel_shape attribute.")
```

**Review by Cadair — COMMENTED** [review]
Mostly comments on the docstrings. In general in sphinx single backticks are references to things and double are plain verbatim text output. You should only use singles for things which resolve to Python objects.

I need to review the tests which hopefully I will do in a bit.

File: ndcube/tests/test_ndcube.py
Comment (DanRyanIrish) [inline ndcube/tests/test_ndcube.py:538]: Could you move these lines to another test that just tests that the correct exception is raised when the WCS is incompatible?

**Review by DanRyanIrish — CHANGES_REQUESTED** [review]
I think this PR will be good to go once these changes have been implemented.

**Review by Cadair — CHANGES_REQUESTED** [review]
The tests look ok. I think the big part missing from this before we land is some narrative documentation. @DanRyanIrish What do yo think the best approach to that is?

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:679]: Got it

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:639]: I think we should also allow a dict / `fits.Header` instance here as well and turn it into a WCS?

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:693]: We should return the type of ourselves so that when people subclass us they get the class back they are expecting.
```suggestion
        resampled_cube = type(self)(data, wcs=target_wcs, meta=deepcopy(self.meta))
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:633]: From sunpy/sunpy#5470 Albert called this `reproject_to` I kind of like that name, thoughts on changing it?

File: ndcube/ndcube.py
Comment (Cadair) [reply ndcube/ndcube.py:639]:
```
isinstance(target_wcs, collections.abc.Mapping):
    target_wcs = astropy.wcs.WCS(target_wcs)
```

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:19]: Please move this into the reproject method.

File: setup.cfg
Comment (Cadair) [inline setup.cfg:39]: move this into a new `reproject` extra in the block below.

File: ndcube/ndcube.py
Comment (ayshih) [inline ndcube/ndcube.py:718]: I believe these should all be `array_shape` instead of `pixel_shape`

File: ndcube/ndcube.py
Comment (ayshih) [inline ndcube/ndcube.py:672]:
```suggestion
            The order of the interpolation. This can be any of: 'nearest-neighbor', 'bilinear',
```
Presumably?

File: ndcube/ndcube.py
Comment (ayshih) [inline ndcube/ndcube.py:680]:
```suggestion
            Whether to return the footprint in addition to the output NDCube.
```

File: ndcube/ndcube.py
Comment (ayshih) [inline ndcube/ndcube.py:669]: You should specify the ordering of dimensions needs to be NumPy ordering

File: ndcube/ndcube.py
Comment (ayshih) [inline ndcube/ndcube.py:665]:
```suggestion
            The WCS object to which the ``NDCube`` is to be reprojected.
```

**Review by ayshih — COMMENTED** [review]
Also, don't raise the generic `Exception`, but instead raise `ValueError` or the like.

**Comment by DanRyanIrish** [comment]
Thanks for these comments @ayshih.  @adwaitbhope, we should definitely address these before merging.

File: changelog/434.feature.rst
Comment (DanRyanIrish) [inline changelog/434.feature.rst:1]:
```suggestion
Adds functionality to reproject `NDCube` to coordinates described by another WCS or FITS Header.
```

File: ndcube/ndcube.py
Comment (DanRyanIrish) [inline ndcube/ndcube.py:705]:
```suggestion
            target_wcs = WCS(header=target_wcs)
```
I don't think a `try...except``` is needed here.  If the header is invalid, `WCS` will raise an error with a more informative traceback.

File: ndcube/ndcube.py
Comment (DanRyanIrish) [inline ndcube/ndcube.py:708]:
```suggestion
            raise ValueError('Given target_wcs is not compatible with this NDCube, the physical types do not match.')
```

File: ndcube/ndcube.py
Comment (DanRyanIrish) [inline ndcube/ndcube.py:717]:
```suggestion
                raise ValueError("shape_out must be specified if target_wcs's low level API "
```

**Review by DanRyanIrish — CHANGES_REQUESTED** [review]

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:718]: Yes, looking back at it, I think it should be `array_shape`. Will change it.

File: ndcube/ndcube.py
Comment (adwaitbhope) [reply ndcube/ndcube.py:705]: Sure

File: ndcube/ndcube.py
Comment (adwaitbhope) [inline ndcube/ndcube.py:663]: Should we use a better name for `target_wcs` now that we're accepting `fits.Header`? Maybe `target_projection` or something.

Comment (DanRyanIrish) [reply ndcube/ndcube.py:663]: I think it's fine.

**Review by DanRyanIrish — APPROVED** [review]

**Comment by DanRyanIrish** [comment]
@Cadair, do think this is ready to merge now?

File: ndcube/ndcube.py
Comment (Cadair) [inline ndcube/ndcube.py:717]:
```suggestion
                raise ValueError("shape_out must be specified if target_wcs does not have the array_shape attribute.")
```

**Review by Cadair — APPROVED** [review]

**Pull request merged**
